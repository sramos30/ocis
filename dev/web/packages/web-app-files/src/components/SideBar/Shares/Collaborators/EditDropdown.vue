<template>
  <span class="oc-flex oc-flex-middle">
    <oc-button :id="editShareBtnId" class="collaborator-edit-dropdown-options-btn" appearance="raw">
      <oc-icon name="more-2" />
    </oc-button>
    <oc-drop
      ref="expirationDateDrop"
      :toggle="'#' + editShareBtnId"
      mode="click"
      padding-size="remove"
    >
      <oc-list class="collaborator-edit-dropdown-options-list" :aria-label="shareEditOptions">
        <li v-if="isExpirationSupported" class="oc-px-s oc-py-xs">
          <oc-datepicker
            v-model="enteredExpirationDate"
            :min-date="minExpirationDate"
            :max-date="maxExpirationDate"
            :locale="$language.current"
            :is-required="isExpirationDateEnforced"
            class="files-recipient-expiration-datepicker"
            data-testid="recipient-datepicker"
          >
            <template #default="{ togglePopover }">
              <oc-button
                class="files-collaborators-expiration-button"
                data-testid="recipient-datepicker-btn"
                appearance="raw"
                @click="togglePopover"
              >
                <span v-if="isExpirationDateSet" v-text="$gettext('Edit expiration date')" />
                <span v-else v-text="$gettext('Set expiration date')" />
              </oc-button>
            </template>
          </oc-datepicker>
        </li>
        <li v-for="(option, i) in options" :key="i" class="oc-px-s oc-py-xs">
          <oc-button
            appearance="raw"
            :class="option.class"
            v-bind="option.additionalAttributes || {}"
            @click="option.method()"
          >
            <span v-text="option.title" />
          </oc-button>
        </li>
      </oc-list>
    </oc-drop>
  </span>
</template>

<script>
import { mapGetters } from 'vuex'
import { DateTime } from 'luxon'
import Mixins from '../../../../mixins'

export default {
  name: 'EditDropdown',
  mixins: [Mixins],
  props: {
    expirationDate: {
      type: Date,
      required: false,
      default: undefined
    },
    shareCategory: {
      type: String,
      required: false,
      default: null,
      validator: function (value) {
        return ['user', 'group'].includes(value) || !value
      }
    }
  },
  data: function () {
    return {
      enteredExpirationDate: null
    }
  },
  computed: {
    ...mapGetters(['capabilities']),

    options() {
      const result = []
      if (this.isRemoveExpirationPossible) {
        result.push({
          title: this.$gettext('Remove expiration date'),
          method: this.removeExpirationDate,
          class: 'remove-expiration-date',
          additionalAttributes: {
            'data-testid': 'collaborator-remove-expiration-btn'
          }
        })
      }
      return [
        ...result,
        {
          title: this.$gettext('Remove share'),
          method: this.removeShare,
          class: 'remove-share',
          additionalAttributes: {
            variation: 'danger',
            'data-testid': 'collaborator-remove-share-btn'
          }
        }
      ]
    },

    editShareBtnId() {
      return 'files-collaborators-edit-button-' + this._uid
    },
    shareEditOptions() {
      return this.$gettext('Context menu of the share')
    },

    editingUser() {
      return this.shareCategory === 'user'
    },

    editingGroup() {
      return this.shareCategory === 'group'
    },

    isExpirationSupported() {
      return (
        (this.editingUser && this.userExpirationDate) ||
        (this.editingGroup && this.groupExpirationDate)
      )
    },

    isExpirationDateSet() {
      return !!this.expirationDate
    },

    isRemoveExpirationPossible() {
      return (
        this.isExpirationSupported && this.isExpirationDateSet && !this.isExpirationDateEnforced
      )
    },

    isDefaultExpirationEnabled() {
      if (this.editingUser) {
        return this.userExpirationDate.enabled
      }

      if (this.editingGroup) {
        return this.groupExpirationDate.enabled
      }

      return this.userExpirationDate.enabled || this.groupExpirationDate.enabled
    },

    userExpirationDate() {
      return this.capabilities.files_sharing.user.expire_date
    },

    groupExpirationDate() {
      return this.capabilities.files_sharing.group?.expire_date
    },

    defaultExpirationDate() {
      if (!this.isDefaultExpirationEnabled) {
        return null
      }

      const userMaxExpirationDays = parseInt(this.userExpirationDate.days)
      const groupMaxExpirationDays = parseInt(this.groupExpirationDate.days)
      let days = 0

      if (this.editingUser) {
        days = userMaxExpirationDays
      } else if (this.editingGroup) {
        days = groupMaxExpirationDays
      } else if (userMaxExpirationDays && groupMaxExpirationDays) {
        days = Math.min(userMaxExpirationDays, groupMaxExpirationDays)
      } else {
        days = userMaxExpirationDays || groupMaxExpirationDays
      }

      const date = new Date()
      date.setDate(new Date().getDate() + days)
      return date
    },

    isExpirationDateEnforced() {
      if (this.editingUser) {
        return this.userExpirationDate.enforced
      }

      if (this.editingGroup) {
        return this.groupExpirationDate.enforced
      }

      return this.userExpirationDate.enforced || this.groupExpirationDate.enforced
    },

    maxExpirationDate() {
      if (!this.isExpirationDateEnforced) {
        return null
      }
      return this.defaultExpirationDate
    },

    minExpirationDate() {
      const date = new Date()
      date.setDate(new Date().getDate() + 1)
      return date
    },

    relativeExpirationDate() {
      return DateTime.fromJSDate(this.enteredExpirationDate)
        .setLocale(this.$language.current)
        .endOf('day')
        .toRelative()
    }
  },
  watch: {
    enteredExpirationDate: {
      handler: 'updateExpirationDate'
    }
  },
  methods: {
    updateExpirationDate() {
      this.$emit('expirationDateChanged', {
        expirationDate: DateTime.fromJSDate(this.enteredExpirationDate).endOf('day').toISO()
      })
      this.$refs.expirationDateDrop.hide()
    },
    removeExpirationDate() {
      this.$emit('expirationDateChanged', {
        expirationDate: null
      })
      this.$refs.expirationDateDrop.hide()
    },
    removeShare() {
      this.$emit('removeShare')
    }
  }
}
</script>
