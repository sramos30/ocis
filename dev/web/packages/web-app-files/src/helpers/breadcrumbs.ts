import { bus } from 'web-pkg/src/instance'

export interface BreadcrumbItem {
  text: string
  to?: string
  allowContextActions?: boolean
  onClick?: () => void
}

export const breadcrumbsFromPath = (
  currentPath: string,
  resourcePath: string
): BreadcrumbItem[] => {
  const pathSplit = (p = '') => p.split('/').filter(Boolean)
  const current = pathSplit(currentPath)
  const resource = pathSplit(resourcePath)

  return resource.map(
    (text, i) =>
      ({
        allowContextActions: true,
        text,
        to: '/' + [...current].splice(0, current.length - resource.length + i + 1).join('/')
      } as BreadcrumbItem)
  )
}

export const concatBreadcrumbs = (...items: BreadcrumbItem[]): BreadcrumbItem[] => {
  const last = items.pop()

  return [
    ...items,
    {
      allowContextActions: last.allowContextActions,
      text: last.text,
      onClick: () => bus.publish('app.files.list.load')
    }
  ]
}
