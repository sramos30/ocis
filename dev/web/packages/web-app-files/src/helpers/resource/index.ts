export * from './asset'
export * from './filter'
export * from './privatePreviewBlob'
export * from './publicPreviewUrl'
export * from './resource'
export * from './sameResource'
