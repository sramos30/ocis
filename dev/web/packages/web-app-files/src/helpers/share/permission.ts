// dummy to trick gettext string extraction into recognizing strings
const $gettext = (str) => {
  return str
}

export class SharePermission {
  private readonly _key: string
  private readonly _bit: number
  private readonly _label: string

  constructor(key: string, bit: number, label: string) {
    this._key = key
    this._bit = bit
    this._label = label
  }

  get key(): string {
    return this._key
  }

  get bit(): number {
    return this._bit
  }

  get label(): string {
    return $gettext(this._label)
  }

  public enabled(bitmask: number): boolean {
    return (bitmask & this.bit) > 0
  }
}

export abstract class SharePermissions {
  static readonly read = new SharePermission('read', 1, $gettext('Read'))
  static readonly update = new SharePermission('update', 2, $gettext('Update'))
  static readonly create = new SharePermission('create', 4, $gettext('Create'))
  static readonly delete = new SharePermission('delete', 8, $gettext('Delete'))
  static readonly share = new SharePermission('share', 16, $gettext('Share'))

  static permissionsToBitmask(permissions: SharePermission[]): number {
    return (permissions || []).reduce((b: number, p: SharePermission) => b | p.bit, 0)
  }

  static bitmaskToPermissions(bitmask: number): SharePermission[] {
    return [this.read, this.update, this.create, this.delete, this.share].reduce(
      (a: SharePermission[], p: SharePermission) => {
        if (p.enabled(bitmask)) {
          a.push(p)
        }
        return a
      },
      []
    )
  }
}
