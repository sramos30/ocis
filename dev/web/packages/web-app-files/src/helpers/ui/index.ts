export * as fileList from './filesList'
export * as resourceTable from './resourceTable'
