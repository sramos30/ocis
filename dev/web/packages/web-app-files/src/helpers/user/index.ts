export * from './avatarUrl'
export * from './types'
