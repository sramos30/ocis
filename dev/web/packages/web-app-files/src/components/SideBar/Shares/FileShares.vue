<template>
  <div id="oc-files-sharing-sidebar" class="oc-position-relative">
    <oc-loader v-if="sharesLoading" :aria-label="$gettext('Loading people list')" />
    <template v-else>
      <invite-collaborator-form v-if="currentUserCanShare" key="new-collaborator" class="oc-my-s" />
      <p
        v-else
        key="no-reshare-permissions-message"
        data-testid="files-collaborators-no-reshare-permissions-message"
        v-text="noResharePermsMessage"
      />
      <div v-if="hasSharees" class="avatars-wrapper oc-flex oc-flex-middle oc-flex-between">
        <h4 class="oc-text-initial oc-text-bold oc-my-rm" v-text="sharedWithLabel" />
        <oc-button
          v-oc-tooltip="sharedWithTooltip"
          data-testid="collaborators-show-people"
          appearance="raw"
          :aria-label="sharedWithTooltip"
          @click="toggleShareeList"
        >
          <oc-avatars
            v-if="!showShareesList"
            :items="collaboratorsAvatar"
            :stacked="true"
            :is-tooltip-displayed="false"
            class="sharee-avatars"
          />
          <oc-icon v-else name="arrow-up-s" />
        </oc-button>
      </div>
      <template v-if="showShareesList && hasSharees">
        <ul
          id="files-collaborators-list"
          class="oc-list oc-list-divider oc-overflow-hidden"
          :class="{ 'oc-mb-l': showSpaceMembers, 'oc-m-rm': !showSpaceMembers }"
          :aria-label="$gettext('Share receivers')"
        >
          <li v-for="collaborator in collaborators" :key="collaborator.key">
            <collaborator-list-item
              :share="collaborator"
              :modifiable="!collaborator.indirect"
              :shared-parent-route="getSharedParentRoute(collaborator)"
              @onDelete="$_ocCollaborators_deleteShare"
            />
          </li>
        </ul>
      </template>
      <template v-if="showSpaceMembers">
        <h4 class="oc-text-initial oc-text-bold oc-my-s" v-text="spaceMemberLabel" />
        <ul
          id="space-collaborators-list"
          class="oc-list oc-list-divider oc-overflow-hidden oc-m-rm"
          :aria-label="spaceMemberLabel"
        >
          <li v-for="collaborator in spaceMembers" :key="collaborator.key">
            <collaborator-list-item :share="collaborator" :modifiable="false" />
          </li>
        </ul>
      </template>
    </template>
  </div>
</template>

<script lang="ts">
import { mapGetters, mapActions, mapState } from 'vuex'
import { watch, computed } from '@vue/composition-api'
import { useStore, useDebouncedRef } from 'web-pkg/src/composables'
import { textUtils } from '../../../helpers/textUtils'
import { getParentPaths } from '../../../helpers/path'
import { dirname } from 'path'
import InviteCollaboratorForm from './InviteCollaborator/InviteCollaboratorForm.vue'
import CollaboratorListItem from './Collaborators/ListItem.vue'
import { ShareTypes } from '../../../helpers/share'
import { clientService } from 'web-pkg/src/services'
import { useTask } from 'vue-concurrency'
import { buildSpace, buildSpaceShare } from '../../../helpers/resources'
import { sortSpaceMembers } from '../../../helpers/space'
import { createLocationSpaces, isLocationSpacesActive } from '../../../router'

export default {
  name: 'FileShares',
  components: {
    InviteCollaboratorForm,
    CollaboratorListItem
  },
  setup() {
    const store = useStore()
    const currentFileOutgoingSharesLoading = computed(
      () => store.getters['Files/currentFileOutgoingSharesLoading']
    )
    const incomingSharesLoading = computed(() => store.state.Files.incomingSharesLoading)
    const sharesTreeLoading = computed(() => store.state.Files.sharesTreeLoading)
    const sharesLoading = useDebouncedRef(true, 250)
    watch([currentFileOutgoingSharesLoading, incomingSharesLoading, sharesTreeLoading], () => {
      sharesLoading.value =
        currentFileOutgoingSharesLoading.value ||
        incomingSharesLoading.value ||
        sharesTreeLoading.value
    })

    const graphClient = clientService.graphAuthenticated(
      store.getters.configuration.server,
      store.getters.getToken
    )

    const loadSpaceTask = useTask(function* (signal, ref, storageId) {
      const graphResponse = yield graphClient.drives.getDrive(storageId)

      if (!graphResponse.data) {
        return
      }

      ref.currentSpace = buildSpace(graphResponse.data)
    })

    const loadSpaceMembersTask = useTask(function* (signal, ref) {
      const promises = []
      const spaceShares = []

      for (const role of Object.keys(ref.currentSpace.spaceRoles)) {
        for (const userId of ref.currentSpace.spaceRoles[role]) {
          promises.push(
            graphClient.users.getUser(userId).then((resolved) => {
              spaceShares.push(buildSpaceShare({ ...resolved.data, role }, ref.currentSpace.id))
            })
          )
        }
      }

      yield Promise.all(promises).then(() => {
        ref.spaceMembers = sortSpaceMembers(spaceShares)
      })
    })

    return { sharesLoading, loadSpaceTask, loadSpaceMembersTask }
  },
  data() {
    return {
      currentShare: null,
      showShareesList: true,
      currentSpace: null,
      spaceMembers: []
    }
  },
  computed: {
    ...mapGetters('Files', ['highlightedFile', 'currentFileOutgoingCollaborators']),
    ...mapState('Files', ['incomingShares', 'sharesTree']),
    ...mapState(['user']),

    sharedWithLabel() {
      return this.$gettext('Shared with')
    },
    spaceMemberLabel() {
      return this.$gettext('Space members')
    },

    hasSharees() {
      return this.collaboratorsAvatar.length > 0
    },

    sharedWithTooltip() {
      return this.showShareesList
        ? this.$gettext('Collapse list of invited people')
        : this.$gettext('Show all invited people')
    },

    collaboratorsAvatar() {
      return this.collaborators.map((c) => {
        return {
          ...c.collaborator,
          shareType: c.shareType
        }
      })
    },

    /**
     * Returns all incoming shares, direct and indirect
     *
     * @return {Array.<Object>} list of incoming shares
     */
    $_allIncomingShares() {
      // direct incoming shares
      const allShares = [...this.incomingShares]

      // indirect incoming shares
      const parentPaths = getParentPaths(this.highlightedFile.path, true)
      if (parentPaths.length === 0) {
        return []
      }

      // remove root entry
      parentPaths.pop()

      parentPaths.forEach((parentPath) => {
        const shares = this.sharesTree[parentPath]
        if (shares) {
          shares.forEach((share) => {
            if (share.incoming) {
              allShares.push(share)
            }
          })
        }
      })

      return allShares
    },

    collaborators() {
      return [...this.currentFileOutgoingCollaborators, ...this.indirectOutgoingShares]
        .sort(this.collaboratorsComparator)
        .map((collaborator) => {
          collaborator.key = 'collaborator-' + collaborator.id
          if (
            collaborator.owner.name !== collaborator.fileOwner.name &&
            collaborator.owner.name !== this.user.id
          ) {
            collaborator.resharers = [collaborator.owner]
          }
          return collaborator
        })
    },

    indirectOutgoingShares() {
      const allShares = []
      const parentPaths = getParentPaths(this.highlightedFile.path, false)
      if (parentPaths.length === 0) {
        return []
      }

      // remove root entry
      parentPaths.pop()

      parentPaths.forEach((parentPath) => {
        const shares = this.sharesTree[parentPath]
        if (shares) {
          shares.forEach((share) => {
            if (share.outgoing && this.$_isCollaboratorShare(share)) {
              share.key = 'indirect-collaborator-' + share.id
              allShares.push(share)
            }
          })
        }
      })

      return allShares
    },

    currentUserCanShare() {
      return this.highlightedFile.canShare({ user: this.user })
    },
    noResharePermsMessage() {
      const translatedFile = this.$gettext("You don't have permission to share this file.")
      const translatedFolder = this.$gettext("You don't have permission to share this folder.")
      return this.highlightedFile.type === 'file' ? translatedFile : translatedFolder
    },

    currentUsersPermissions() {
      if (this.$_allIncomingShares.length > 0) {
        let permissions = 0

        for (const share of this.$_allIncomingShares) {
          permissions |= share.permissions
        }

        return permissions
      }

      return null
    },
    currentUserIsMemberOfSpace() {
      return this.currentSpace?.spaceMemberIds.includes(this.user.uuid)
    },
    showSpaceMembers() {
      return (
        this.currentSpace &&
        this.highlightedFile.type !== 'space' &&
        this.currentUserIsMemberOfSpace
      )
    }
  },
  watch: {
    highlightedFile: {
      handler: function (newItem, oldItem) {
        if (oldItem !== newItem) {
          this.$_reloadShares()
          this.showShareesList = true
        }
      },
      immediate: true
    }
  },
  async mounted() {
    if (this.$route.params.storageId) {
      await this.loadSpaceTask.perform(this, this.$route.params.storageId)

      if (this.showSpaceMembers) {
        this.loadSpaceMembersTask.perform(this)
      }
    }
  },
  methods: {
    ...mapActions('Files', [
      'loadCurrentFileOutgoingShares',
      'loadSharesTree',
      'deleteShare',
      'loadIncomingShares'
    ]),
    $_isCollaboratorShare(collaborator) {
      return ShareTypes.containsAnyValue(ShareTypes.authenticated, [collaborator.shareType])
    },
    collaboratorsComparator(c1, c2) {
      // Sorted by: type, direct, display name, creation date
      const c1DisplayName = c1.collaborator ? c1.collaborator.displayName : c1.displayName
      const c2DisplayName = c2.collaborator ? c2.collaborator.displayName : c2.displayName
      const name1 = c1DisplayName.toLowerCase().trim()
      const name2 = c2DisplayName.toLowerCase().trim()
      const c1UserShare = ShareTypes.containsAnyValue(ShareTypes.individuals, [c1.shareType])
      const c2UserShare = ShareTypes.containsAnyValue(ShareTypes.individuals, [c2.shareType])
      const c1DirectShare = !c1.indirect
      const c2DirectShare = !c2.indirect

      if (c1UserShare === c2UserShare) {
        if (c1DirectShare === c2DirectShare) {
          if (name1 === name2) {
            return textUtils.naturalSortCompare(c1.stime + '', c2.stime + '')
          }

          return textUtils.naturalSortCompare(name1, name2)
        }

        return c1DirectShare ? -1 : 1
      }

      return c1UserShare ? -1 : 1
    },
    toggleShareeList() {
      this.showShareesList = !this.showShareesList
    },
    $_ocCollaborators_deleteShare(share) {
      this.deleteShare({
        client: this.$client,
        share: share,
        resource: this.highlightedFile,
        storageId: this.$route.params.storageId
      })
    },
    $_reloadShares() {
      this.loadCurrentFileOutgoingShares({
        client: this.$client,
        path: this.highlightedFile.path,
        $gettext: this.$gettext,
        storageId: this.$route.params.storageId
      })
      this.loadIncomingShares({
        client: this.$client,
        path: this.highlightedFile.path,
        $gettext: this.$gettext,
        storageId: this.$route.params.storageId
      })
      this.loadSharesTree({
        client: this.$client,
        path: dirname(this.highlightedFile.path),
        $gettext: this.$gettext,
        storageId: this.$route.params.storageId
      })
    },
    getSharedParentRoute(parentShare) {
      if (!parentShare.indirect) {
        return null
      }

      if (this.sharesTree[parentShare.path]) {
        if (isLocationSpacesActive(this.$router, 'files-spaces-project')) {
          return createLocationSpaces('files-spaces-project', {
            params: { storageId: this.$route.params.storageId, item: parentShare.path }
          })
        }

        return createLocationSpaces('files-spaces-personal-home', {
          params: { item: parentShare.path }
        })
      }

      return null
    }
  }
}
</script>

<style>
.avatars-wrapper {
  height: 40px;
}
</style>
