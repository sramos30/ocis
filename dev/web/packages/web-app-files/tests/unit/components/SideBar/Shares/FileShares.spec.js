import FileShares from '@files/src/components/SideBar/Shares/FileShares.vue'
import { createLocalVue, mount, shallowMount } from '@vue/test-utils'
import GetTextPlugin from 'vue-gettext'
import Vuex from 'vuex'
import DesignSystem from 'owncloud-design-system'
import Users from '@/__fixtures__/users'
import Collaborators from '@/__fixtures__/collaborators'
import mockAxios from 'jest-mock-axios'
import { spaceRoleManager } from '../../../../../src/helpers/share'
import VueCompositionAPI from '@vue/composition-api/dist/vue-composition-api'

const localVue = createLocalVue()
localVue.use(DesignSystem)
localVue.use(Vuex)
localVue.use(VueCompositionAPI)
localVue.use(GetTextPlugin, {
  translations: 'does-not-matter.json',
  silent: true
})

const user = Users.alice
const collaborators = [Collaborators[0], Collaborators[1]]

const selectors = {
  showCollaboratorButton: 'button[data-testid="collaborators-show-people"]',
  firstCollaboratorListItem: `div[data-testid="collaborator-user-item-${Collaborators[0].collaborator.name}"]`
}

describe('FileShares', () => {
  it('renders loading spinner while loading', () => {
    const wrapper = getShallowMountedWrapper(
      {
        user
      },
      true
    )
    expect(wrapper).toMatchSnapshot()
  })

  describe('if currentUser can share', () => {
    it('initially renders add people dialog', () => {
      const wrapper = getShallowMountedWrapper({
        user
      })
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('if currentUser can not share', () => {
    it('initially renders no share permission message', () => {
      const wrapper = getShallowMountedWrapper({
        user,
        canShare: false
      })
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('if there are no collaborators', () => {
    it('does not render avatar wrapper or collaborator list', () => {
      const wrapper = getShallowMountedWrapper({
        user
      })
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('if there are collaborators present', () => {
    it('renders sharedWithLabel and sharee list', () => {
      const wrapper = getShallowMountedWrapper({
        user,
        outgoingCollaborators: collaborators
      })
      expect(wrapper).toMatchSnapshot()
    })

    it('can toggle the collaborators list by clicking the avatar wrapper button', async () => {
      const wrapper = getMountedWrapper({
        user,
        outgoingCollaborators: collaborators
      })
      const button = wrapper.find(selectors.showCollaboratorButton)
      expect(wrapper.vm.showShareesList).toBe(true)
      await button.trigger('click')
      expect(wrapper.vm.showShareesList).toBe(false)
      await button.trigger('click')
      expect(wrapper.vm.showShareesList).toBe(true)
    })

    it('reacts on delete events by collaborator list items', async () => {
      const spyOnCollaboratorDelete = jest
        .spyOn(FileShares.methods, '$_ocCollaborators_deleteShare')
        .mockImplementation()
      const wrapper = getMountedWrapper({
        user,
        outgoingCollaborators: collaborators
      })
      wrapper.find(selectors.firstCollaboratorListItem).vm.$emit('onDelete')
      await wrapper.vm.$nextTick()
      expect(spyOnCollaboratorDelete).toHaveBeenCalledTimes(1)
    })
    it('reloads shares if highlighted file is changed', async () => {
      const spyOnReloadShares = jest
        .spyOn(FileShares.methods, '$_reloadShares')
        .mockImplementation()
      const wrapper = getMountedWrapper({
        user,
        outgoingCollaborators: collaborators
      })
      wrapper.vm.$store.commit('Files/SET_HIGHLIGHTED_FILE', { name: 'testfile2' })
      await wrapper.vm.$nextTick()
      expect(spyOnReloadShares).toHaveBeenCalledTimes(1)
    })
    it('correctly passes the shared parent route to the collaborator list item', () => {
      const wrapper = getShallowMountedWrapper({
        user,
        outgoingCollaborators: [{ ...Collaborators[0], indirect: true }]
      })
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('current space', () => {
    afterEach(() => {
      mockAxios.reset()
    })
    it('loads space members if a space is given', async () => {
      const spaceMock = {
        id: '1',
        root: { permissions: [{ roles: ['manager'], grantedTo: [{ user: { id: 1 } }] }] }
      }
      mockAxios.request.mockImplementationOnce(() => {
        return Promise.resolve({
          data: spaceMock
        })
      })
      mockAxios.request.mockImplementationOnce(() => {
        return Promise.resolve({
          data: { role: spaceRoleManager.name }
        })
      })

      const wrapper = getShallowMountedWrapper({ user })

      await wrapper.vm.loadSpaceTask.last
      await wrapper.vm.loadSpaceMembersTask.last
      expect(wrapper.vm.spaceMembers.length).toBe(1)
      expect(wrapper.find('#space-collaborators-list').exists()).toBeTruthy()
    })
    it('does not load space members if no space is given', () => {
      const wrapper = getShallowMountedWrapper({
        user
      })

      expect(wrapper.vm.spaceMembers.length).toBe(0)
    })
  })
})

function getResource({
  filename = 'testFile',
  extension = 'txt',
  type = 'file',
  owner = 'user0',
  canShare = true
}) {
  return {
    id: '4',
    fileId: '4',
    icon: type,
    name: type === 'file' ? `${filename}.${extension}` : filename,
    extension: extension,
    path: type === 'file' ? `/${filename}.${extension}` : `/${filename}`,
    type,
    isFolder: type === 'folder',
    mdate: 'Mon, 12 Jul 2021 11:04:33 GMT',
    size: '163',
    indicators: [],
    permissions: 'RDNVW',
    starred: false,
    etag: '"89128c0e8122002db57bd19c9ec33004"',
    shareTypes: [],
    downloadURL: '',
    ownerDisplayName: owner.displayname,
    ownerId: owner.id,
    canDownload: () => true,
    isReceivedShare: () => true,
    canBeDeleted: () => true,
    canRename: () => true,
    canShare: () => canShare
  }
}

const storeOptions = (data) => {
  let {
    user,
    outgoingCollaborators = [],
    incomingCollaborators = [],
    owner,
    canShare = true
  } = data
  if (!owner) {
    owner = user
  }

  return {
    state: {
      user
    },
    modules: {
      Files: {
        state: {
          incomingShares: incomingCollaborators,
          sharesTree: { [Collaborators[0].path]: [Collaborators[0]] }
        },
        namespaced: true,
        getters: {
          highlightedFile: () => {
            return getResource({ filename: 'testfile', extension: 'jpg', type: 'file', canShare })
          },
          currentFileOutgoingCollaborators: () => outgoingCollaborators,
          currentFileOutgoingSharesLoading: () => false,
          sharesTreeLoading: () => false
        },
        actions: {
          loadCurrentFileOutgoingShares: jest.fn(),
          loadIncomingShares: jest.fn(),
          loadSharesTree: jest.fn(),
          deleteShare: jest.fn()
        },
        mutations: {
          SET_HIGHLIGHTED_FILE(state, file) {
            const newFile = getResource(file)
            state.highlightedFile = newFile
          }
        }
      }
    },
    getters: {
      getToken: jest.fn(() => 'GFwHKXdsMgoFwt'),
      configuration: jest.fn(() => ({
        server: 'http://example.com/'
      })),
      user: () => user,
      capabilities: () => {
        return {
          files_sharing: {
            user: {
              expire_date: {
                enabled: true,
                days: 10
              }
            },
            group: {
              expire_date: {
                enabled: true,
                days: 10
              }
            }
          }
        }
      }
    }
  }
}

function getMountedWrapper(data, loading = false) {
  return mount(FileShares, {
    localVue,
    store: createStore(data),
    mocks: {
      sharesLoading: loading,
      $route: {
        params: { storageId: 1 }
      },
      $router: {
        currentRoute: { name: 'some-route' },
        resolve: (r) => {
          return {
            href: r.name
          }
        }
      }
    },
    stubs: {
      'oc-button': false,
      'oc-icon': true,
      'oc-spinner': true,
      'avatar-image': true
    }
  })
}

function getShallowMountedWrapper(data, loading = false) {
  return shallowMount(FileShares, {
    localVue,
    store: createStore(data),
    stubs: {
      'oc-button': true,
      'oc-icon': true,
      'oc-spinner': true
    },
    mocks: {
      sharesLoading: loading,
      $route: {
        params: {
          storageId: 1
        }
      },
      $router: {
        currentRoute: { name: 'some-route' },
        resolve: (r) => {
          return {
            href: r.name
          }
        }
      }
    }
  })
}

function createStore(data) {
  return new Vuex.Store(storeOptions(data))
}
