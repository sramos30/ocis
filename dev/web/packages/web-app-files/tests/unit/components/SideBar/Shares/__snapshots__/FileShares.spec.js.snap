// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileShares if currentUser can not share initially renders no share permission message 1`] = `
<div id="oc-files-sharing-sidebar" class="oc-position-relative">
  <p data-testid="files-collaborators-no-reshare-permissions-message">You don't have permission to share this file.</p>
  <!---->
  <!---->
  <!---->
</div>
`;

exports[`FileShares if currentUser can share initially renders add people dialog 1`] = `
<div id="oc-files-sharing-sidebar" class="oc-position-relative">
  <invite-collaborator-form-stub class="oc-my-s"></invite-collaborator-form-stub>
  <!---->
  <!---->
  <!---->
</div>
`;

exports[`FileShares if there are collaborators present correctly passes the shared parent route to the collaborator list item 1`] = `
<div id="oc-files-sharing-sidebar" class="oc-position-relative">
  <invite-collaborator-form-stub class="oc-my-s"></invite-collaborator-form-stub>
  <div class="avatars-wrapper oc-flex oc-flex-middle oc-flex-between">
    <h4 class="oc-text-initial oc-text-bold oc-my-rm">Shared with</h4>
    <oc-button-stub type="button" size="medium" arialabel="Collapse list of invited people" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" data-testid="collaborators-show-people">
      <oc-icon-stub name="arrow-up-s" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub>
    </oc-button-stub>
  </div>
  <ul id="files-collaborators-list" aria-label="Share receivers" class="oc-list oc-list-divider oc-overflow-hidden oc-m-rm">
    <li>
      <collaborator-list-item-stub share="[object Object]" sharedparentroute="[object Object]"></collaborator-list-item-stub>
    </li>
  </ul>
  <!---->
</div>
`;

exports[`FileShares if there are collaborators present renders sharedWithLabel and sharee list 1`] = `
<div id="oc-files-sharing-sidebar" class="oc-position-relative">
  <invite-collaborator-form-stub class="oc-my-s"></invite-collaborator-form-stub>
  <div class="avatars-wrapper oc-flex oc-flex-middle oc-flex-between">
    <h4 class="oc-text-initial oc-text-bold oc-my-rm">Shared with</h4>
    <oc-button-stub type="button" size="medium" arialabel="Collapse list of invited people" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" data-testid="collaborators-show-people">
      <oc-icon-stub name="arrow-up-s" filltype="fill" accessiblelabel="" type="span" size="medium" variation="passive" color=""></oc-icon-stub>
    </oc-button-stub>
  </div>
  <ul id="files-collaborators-list" aria-label="Share receivers" class="oc-list oc-list-divider oc-overflow-hidden oc-m-rm">
    <li>
      <collaborator-list-item-stub share="[object Object]" modifiable="true"></collaborator-list-item-stub>
    </li>
    <li>
      <collaborator-list-item-stub share="[object Object]" modifiable="true"></collaborator-list-item-stub>
    </li>
  </ul>
  <!---->
</div>
`;

exports[`FileShares if there are no collaborators does not render avatar wrapper or collaborator list 1`] = `
<div id="oc-files-sharing-sidebar" class="oc-position-relative">
  <invite-collaborator-form-stub class="oc-my-s"></invite-collaborator-form-stub>
  <!---->
  <!---->
  <!---->
</div>
`;

exports[`FileShares renders loading spinner while loading 1`] = `
<div id="oc-files-sharing-sidebar" class="oc-position-relative">
  <oc-loader-stub arialabel="Loading people list"></oc-loader-stub>
</div>
`;
